FreeRTOS/Demo/Common/Minimal/TimerDemo.c

```diff
@@ -664,6 +684,75 @@ uint8_t ucTimer;
 }
 /*-----------------------------------------------------------*/
 
+static void prvTest7_CheckBacklogBehaviour( void )
+{
+	/* Use the first auto-reload timer to test stopping a timer from a
+	backlogged callback. */
+
+	/* The timer has not been started yet! */
+	if( xTimerIsTimerActive( xAutoReloadTimers[ 0 ] ) != pdFALSE )
+	{
+		xTestStatus = pdFAIL;
+		configASSERT( xTestStatus );
+	}
+
+	ucAutoReloadTimerCounters[ 0 ] = ( uint8_t ) 0;
+
+	/* Prompt the callback function to stop the timer. */
+	ucIsStopNeededInTimerZeroCallback = ( uint8_t ) pdTRUE;
+
+	/* Now start the timer.  This will appear to happen immediately to
+	this task because this task is running at a priority below the timer
+	service task.  Use a timer period small enough that the call to
+	xTaskCatchUpTicks() below has a small impact on other tests that might
+	be running, but large enough that other tests won't cause an extra
+	(unexpected) timer expiry. */
+	#define tmrdemoBACKLOG_TIMER_PERIOD		( ( TickType_t ) 5 )
+	xTimerChangePeriod( xAutoReloadTimers[ 0 ], tmrdemoBACKLOG_TIMER_PERIOD, tmrdemoDONT_BLOCK );
+
+	/* The timer should now be active. */
+	if( xTimerIsTimerActive( xAutoReloadTimers[ 0 ] ) == pdFALSE )
+	{
+		xTestStatus = pdFAIL;
+		configASSERT( xTestStatus );
+	}
+
+	/* Arrange for the callback to execute late enough that it will execute
+	twice, back-to-back.  The timer must handle the stop request properly
+	in spite of the backlog of callbacks. */
+	#define tmrdemoEXPECTED_BACKLOG_EXPIRIES  ( ( TickType_t ) 2 )
+	xTaskCatchUpTicks( tmrdemoBACKLOG_TIMER_PERIOD * tmrdemoEXPECTED_BACKLOG_EXPIRIES );
+
+	/* The timer should now be inactive. */
+	if( xTimerIsTimerActive( xAutoReloadTimers[ 0 ] ) != pdFALSE )
+	{
+		xTestStatus = pdFAIL;
+		configASSERT( xTestStatus );
+	}
+
+	/* Verify the backlog cleared properly. */
+	if( ucAutoReloadTimerCounters[ 0 ] != ( uint8_t ) tmrdemoEXPECTED_BACKLOG_EXPIRIES )
+	{
+		xTestStatus = pdFAIL;
+		configASSERT( xTestStatus );
+	}

```

Comment:
On second thought, before we enshrine this check as a requirement and have some user start depending on it, maybe we shouldn't.

What's the harm if the timer task processes the stop command and decides not to work through the entire backlog?  Callers already accept ambiguity in *when* timer commands are processed.  So perhaps it's better to detect an error here with `>` instead of `!=`, verifying only that the stop command didn't get processed too late.  

Or even better eliminate this check altogether.  That would allow us to reduce `tmrdemoBACKLOG_TIMER_PERIOD` to 1 and make this test less disruptive to other tests because the value passed to `xTaskCatchUpTicks()` would be very small (2).

Either way a change here will slightly simplify and reduce https://github.com/FreeRTOS/FreeRTOS-Kernel/pull/305.